### PR TITLE
cmake: Add dpdk numa support

### PR DESCRIPTION
**Pull request by rosinL**
Find and link the numa library for DPDK.

Signed-off-by: Hu Ye <yehu5@huawei.com>
Signed-off-by: Chunsong Feng <fengchunsong@huawei.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/msg/CMakeLists.txt [hunk 1]
```diff
@@ -57,6 +57,10 @@ if(WITH_DPDK)
     ${async_dpdk_srcs})
   target_link_libraries(common_async_dpdk PRIVATE
     dpdk::dpdk)
+  find_package(numa QUIET)
```
Comment (tchaikov) [inline src/msg/CMakeLists.txt]: can we assume that libnuma is linked explicitly only if libdpdk is built from the submodule? if that's the case, probably we can move the numa stuff to `BuildDPDK.cmake`? and add this linkage to `dpdk::eal` instead ?

File: cmake/modules/patch-dpdk-conf.sh [hunk 2]
```diff
@@ -50,6 +50,3 @@ setconf CONFIG_RTE_TEST_PMD n
 
 # async/dpdk does not like it
 setconf CONFIG_RTE_MBUF_REFCNT_ATOMIC n
-
-# balanced allocation of hugepages
-setconf CONFIG_RTE_EAL_NUMA_AWARE_HUGEPAGES n
```
Comment (tchaikov) [inline cmake/modules/patch-dpdk-conf.sh]: this breaks the build if DPDK enables `CONFIG_RTE_EAL_NUMA_AWARE_HUGEPAGES` by default, while libnuma-dev/libnuma-devel is not installed.

File: cmake/modules/BuildDPDK.cmake [hunk 3]
```diff
@@ -95,8 +95,13 @@ function(do_build_dpdk dpdk_dir)
     BUILD_COMMAND ${make_cmd} O=${dpdk_dir} CC=${CMAKE_C_COMPILER} EXTRA_CFLAGS=-fPIC
     BUILD_IN_SOURCE 1
     INSTALL_COMMAND "true")
+  if(numa_FOUND)
```
Comment (tchaikov) [inline cmake/modules/BuildDPDK.cmake]: could you move `find_package(numa QUIET)` here ? to where `numa_FOUND` is used the first time.

File: cmake/modules/Findnuma.cmake [hunk 4]
```diff
@@ -0,0 +1,16 @@
+# - Find libnuma
```
Comment (tchaikov) [inline cmake/modules/Findnuma.cmake]: i'd suggest rename `Findnuma.cmake` to `FindNUMA.cmake`. as like ALSA, BLAS, NUMA is also an abbreviation. or, at least, we should be consistent on the naming of exposed variables. like `Findfoo.cmake` => `foo_INCLUDE_DIR` .

File: src/msg/CMakeLists.txt [hunk 5]
```diff
@@ -57,6 +57,10 @@ if(WITH_DPDK)
...
+  if(numa_FOUND)
+    target_link_libraries(common_async_dpdk INTERFACE numa)
+  endif()
```
Comment (tchaikov) [inline src/msg/CMakeLists.txt]: is this still needed?

Comment (fengchunsong) [reply src/msg/CMakeLists.txt]: No longer needed, thanks for correction

File: cmake/modules/BuildDPDK.cmake (hunk 3 above)
Comment (fengchunsong) [reply cmake/modules/BuildDPDK.cmake]: The result of find_package (numa QUIET)  is used in both do_build_dpdk and do_export_dpdk. Put find_package (numa QUIET) in do_build_dpdk, NUMA_FOUND is true in do_build_dpdk but false in do_export_dpdk.

File: cmake/modules/BuildDPDK.cmake [hunk 6]
```diff
@@ -161,21 +166,25 @@ function(do_export_dpdk dpdk_dir)
...
+  endif()
   add_library(dpdk::dpdk INTERFACE IMPORTED)
   add_dependencies(dpdk::dpdk
     ${DPDK_LIBRARIES})
   # workaround for https://gitlab.kitware.com/cmake/cmake/issues/16947
   set_target_properties(dpdk::dpdk PROPERTIES
     INTERFACE_INCLUDE_DIRECTORIES ${DPDK_INCLUDE_DIR}
     INTERFACE_LINK_LIBRARIES
-    "-Wl,--whole-archive $<JOIN:${DPDK_ARCHIVES}, > -Wl,--no-whole-archive")
+    "-Wl,--whole-archive $<JOIN:${DPDK_ARCHIVES}, > -Wl,--no-whole-archive -Wl,-lpthread,-ldl${dpdk_numa}")
```
Comment (tchaikov) [inline cmake/modules/BuildDPDK.cmake]: why is `-lpthread,-ldl` needed now?

File: src/msg/CMakeLists.txt (hunk 5 above)
Comment (tchaikov) [reply src/msg/CMakeLists.txt]: this comment is not addressed yet.

File: cmake/modules/BuildDPDK.cmake (hunk 6 above)
Comment (fengchunsong) [reply cmake/modules/BuildDPDK.cmake]: librte_eal.a refers to libpthread.so and libceph-common.so refers to librte_eal.a. Many test programs link libceph-common.so, but without linking pthread, it will cause compilation error. Such as unittest_interval_set, target_link_libraries (unittest_interval_set ceph-common)

Comment (tchaikov) [reply cmake/modules/BuildDPDK.cmake]: could you add this in a separated commit?

Comment (fengchunsong) [reply cmake/modules/BuildDPDK.cmake]: ok,thanks

File: cmake/modules/BuildDPDK.cmake [hunk 7]
```diff
@@ -176,7 +176,7 @@ function(do_export_dpdk dpdk_dir)
   set_target_properties(dpdk::dpdk PROPERTIES
     INTERFACE_INCLUDE_DIRECTORIES ${DPDK_INCLUDE_DIR}
     INTERFACE_LINK_LIBRARIES
-    "-Wl,--whole-archive $<JOIN:${DPDK_ARCHIVES}, > -Wl,--no-whole-archive")
+    "-Wl,--whole-archive $<JOIN:${DPDK_ARCHIVES}, > -Wl,--no-whole-archive ${dpdk_numa} -Wl,-lpthread,-ldl")
```
Comment (tchaikov) [inline cmake/modules/BuildDPDK.cmake]: `${dpdk_numa}` does not belong to this commit.

Comment (fengchunsong) [reply cmake/modules/BuildDPDK.cmake]: ${dpdk_numa} belongs to commit 58d0be5dcee41b19f8d94654c3b05177cd3b2495.
commit 98229bf69a8cb22af70504feeec448a99ee68abd only add '-Wl,-lpthread,-ldl',don't contain any numa

Comment (rosinL) [reply cmake/modules/BuildDPDK.cmake]: ok,I split the commit

Comment (fengchunsong) [reply cmake/modules/BuildDPDK.cmake]: 98229bf  depends on 58d0be5.

File: cmake/modules/patch-dpdk-conf.sh [hunk 8]
```diff
@@ -17,6 +17,8 @@ machine=$1
 shift
 arch=$1
 shift
+numa=$1
```
Comment (tchaikov) [inline cmake/modules/patch-dpdk-conf.sh]: we will need a better way to pass the parameters if the list gets longer over time.

**Comment by tchaikov** [comment]
```
9/190 Test  #15: test_objectstore_memstore.sh ............***Failed   28.66 sec
```

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2019-11-30_02:34:26-rados-wip-kefu-testing-2019-11-29-2231-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2019-12-02_01:50:54-rados-wip-kefu-testing-2019-11-29-2231-distro-basic-mira/

failures tracked by:

- https://tracker.ceph.com/issues/42764
- https://tracker.ceph.com/issues/43075
- https://tracker.ceph.com/issues/43077
- https://tracker.ceph.com/issues/43078
